### ISSUE TITLE
VRAM requirements/benchmarks?

### ISSUE THREAD

**Issue by Zarxrax** [facebookresearch/sam2#118]
Is there any data available on how much VRAM is required to run the model at various resolutions? Does the requirement increase depending on the length of the video? I did not find any information about this in the paper.

**Comment by heyoeyo** [comment]
I haven't gotten to the video stuff yet, but here are some values for peak VRAM usage (from `torch.cuda.max_memory_allocated()`) when processing a single image for 3 of the models:

**Tiny model** (small model is very similar)
| Resolution | Peak VRAM (MB) |
| -----------| ----------|
| 576x1024 | 240 |
| 1024x1024   | 458 | Tiny |
| 2048x2048 | 1158 | Tiny |

**Base-plus model**
| Resolution | Peak VRAM (MB) |
| -----------| ----------|
| 576x1024 | 328 | 
| 1024x1024 | 551 | 
| 2048x2048 | 1275 |

**Large model**
| Resolution | Peak VRAM (MB) |
| -----------| ----------|
| 576x1024 | 624 |
| 1024x1024 | 855 |
| 2048x2048 | 1626 |

This is all running with float16 and without any of the [fancy attention optimizations](https://github.com/facebookresearch/segment-anything-2/blob/57bc94b7391e47e5968004a0698f8bf793a544d1/sam2/modeling/sam/transformer.py#L249-L254).

I haven't checked out the video processing yet, but I'd expect that the image encoder makes up the bulk of the VRAM usage, so I'd be surprised if the memory requirements were much higher (unless processing videos accumulates data over time, I'm not sure!).

**Comment by Zarxrax** [comment]
Thanks, that doesn't look bad at all for single image processing.

I do think on video it will probably use significantly more VRAM though (that's what happens on other video segmentation models), though it would be awesome if they have made progress on that front.

**Comment by heyoeyo** [comment]
Yes I think you're right. Having a glance at the video code, it looks like the entire frame sequence along with per-frame [image encodings are cached](https://github.com/facebookresearch/segment-anything-2/blob/57bc94b7391e47e5968004a0698f8bf793a544d1/sam2/sam2_video_predictor.py#L741), among other things.

It looks like it is possible to [offload some of the memory usage](https://github.com/facebookresearch/segment-anything-2/blob/57bc94b7391e47e5968004a0698f8bf793a544d1/sam2/sam2_video_predictor.py#L42-L43) to regular RAM, but this doesn't seem to apply to the image encoding, and regardless it's still a build-up... may be a problem for long videos!

**Comment by heyoeyo** [comment]
Just as a follow up for anyone finding this, I did eventually get around to testing the video processing.
The original code caches results _indefinitely_, leading to a continual increase in VRAM usage over time.

However, this isn't necessary (or even particularly useful), and the model actually does have a finite VRAM requirement if the caching is disabled:

| Model | VRAM |
| ----- | ---- |
| Tiny | 1.2 GB |
| Base | 1.3 GB |
| Large | 1.7 GB |

These are the steady-state values using a 1024x1024 processing resolution, tracking a single object. Every additional object adds something like 3-4MB.
These number aren't exact (they come from calling `torch.cuda.get_mem_info()`), but the main takeaway is that even the largest model can track indefinitely with <2GB of VRAM if the internal caching is disabled.
